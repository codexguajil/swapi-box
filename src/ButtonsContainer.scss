#buttons-container {
    float: right;
    margin-left: 250px;
    margin-top: 70px;
    width: 80%;
}

button {
    margin: 15px;
}