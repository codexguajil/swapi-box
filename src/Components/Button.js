import React, { Component } from 'react';
import propTypes from 'prop-types';

export default class Button extends Component {
    constructor(props) {
        super(props);
        this.state={
            clicked: false,
        }
    }

    handleClick = (e) => {
        this.setState({clicked: !this.state.clicked})
        this.props.fetch(e.target.value)
    }

    render() {
        return (
            <button id={this.state.clicked.toString()}
                    onClick={this.handleClick}
                    value={this.props.value}
            >{this.props.value}
            </button>
        )
    }
}

Button.propTypes = {
    fetch: propTypes.func.isRequired
}