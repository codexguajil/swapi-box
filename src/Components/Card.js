import React, { Component } from 'react';

export default class Card extends Component {

    render() {
      return (
        <div className="card">
          <h1>{this.props.data.name}</h1>
          <h3>{this.props.data.homeworld}</h3>
          <h4>{this.props.data.species}</h4>
          <h4>{this.props.data.population}</h4>
        </div>
      )
    }
  }