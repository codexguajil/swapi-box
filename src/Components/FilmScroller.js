import React from 'react';
import propTypes from 'prop-types';

export const FilmScroller = ({films}) => {

    let randomNum = Math.floor(Math.random() * films.length)

   return (
       <div id="film">
            <div id="crawl">
            <h1>welcome</h1>
            {films.length > 0 &&
                    <h3>
                        {films[randomNum].opening_crawl}
                        <p>{films[randomNum].title}</p>
                        <p>{films[randomNum].release_date}</p>
                    </h3>
            }
        </div>
       </div>
   )
}

FilmScroller.propTypes = {
    films: propTypes.array.isRequired
}