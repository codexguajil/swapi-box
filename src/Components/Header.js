import React from 'react';
import Favorites from '../containers/Favorites/favorites';

export const Header = () => {

    return (
        <div id="header">
            <h1>SWAPI-Box</h1>
            <Favorites />
        </div>
    )
}