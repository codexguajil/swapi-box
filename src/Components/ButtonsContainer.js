import React, { Component } from 'react';
import Button from './Button';
import propTypes from 'prop-types';

export const ButtonsContainer = (props) => {
    
        return (
            <div id="buttons-container">
                <Button value="people" fetch={props.fetchPeople}/>
                <Button value="planets" fetch={props.fetchPeople}/>
                <Button value="vehicles" fetch={props.fetchPeople}/>
            </div>
        )
}

ButtonsContainer.propTypes = {
    fetchPeople: propTypes.func.isRequired
}