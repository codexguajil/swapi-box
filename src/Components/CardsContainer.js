import React from 'react';
import Card from './Card';
import propTypes from 'prop-types';

export const CardsContainer = ({data}) => {
    return (
            <div id="cards-container">
            {data.length > 0 &&
                data.map(data => {
                    return <Card data={data}/>
                })
            }
            </div>
    )
}

CardsContainer.propTypes = {
    data: propTypes.array.isRequired
}