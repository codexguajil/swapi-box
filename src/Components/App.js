import React, { Component } from 'react';
import { FilmScroller } from './FilmScroller';
import { Header } from './Header';
import { CardsContainer } from './CardsContainer';
import { ButtonsContainer } from './ButtonsContainer';
import { fetchFilms, fetchApi, iterateFetch } from '../Methods/apiCalls';
import { iterateHelper } from '../Methods/helpers';
import './App.css';
import { promises } from 'fs';

class App extends Component {
  constructor() {
    super();
    this.state={
      films: [],
      count: 0,
      data: [],
    }
    this.fetchPeople = this.fetchPeople.bind(this)
    this.fetchSpecies = this.fetchSpecies.bind(this)
  }

  componentDidMount() {
    const url = 'https://swapi.co/api/films/'
    this.fetchFilms(url)
  }

  async fetchFilms(url) {
    try {
      const films = await fetchFilms(url)
      this.setState({films, count: films.length})
    } catch(error) {
      console.log(error.message)
    }
  }

  async fetchPeople(value) {  
    console.log(value)
    const url = `https://swapi.co/api/${value}`
    try {
      const parsedResponse = await fetchApi(url)
      const objectValues = await iterateHelper(parsedResponse.results, value)
      console.log('objectVAlues:', objectValues)
      if(objectValues[0].species) {
        let homeWorlds = await this.fetchHomeWorlds(objectValues)
        let species = await this.fetchSpecies(objectValues)
        homeWorlds.forEach((person,i) => {
          person.species = species[i].name
          person.language = species[i].language
        })
        console.log(homeWorlds);
        this.setState({data: homeWorlds})
        } else if(objectValues[0].terrain) {
          let residents = await this.fetchResidents(objectValues)
          let newPlanets = this.resetResidents(objectValues, residents)
          console.log(newPlanets)
          this.setState({data: newPlanets})
        } else if(objectValues[0].model) {
          this.setState({data: objectValues})
        }
      } catch(error) {
        console.log(error.message)
      }
    }

  resetResidents = (planetValues, residents) => {
    let resetPlanetValues = planetValues.map((planet,i) => {
      return {...planet, residents: residents[i]}
    })
    return resetPlanetValues
  }

  async fetchResidents(planets) {
    try {
      let newPlanets = planets.map(planet => {
        let fetchedResidents = this.fetchUrls(planet.residents)
        return fetchedResidents
      })
      return await Promise.all(newPlanets)
    }catch(error) {
      console.log(error.message)
    }
  }

  async fetchUrls(urls) {
    if(urls.length > 0) {
      let fetchedUrls = await urls.map(url => {
        return fetchApi(url)
      })
      return await Promise.all(fetchedUrls)
    } else {
      urls.push({name:'unknown'})
      return urls;
    }
  }

  async fetchHomeWorlds(people) {
    console.log('people', people)
    let homes = people.map(person => {
       return fetchApi(person.home)
    })
    let allHomes = await Promise.all(homes)
    let homeValues = allHomes.map((home,i) => {
      return {
        name: people[i].name,
        Homeworld: home.name,
        Population: home.population
      }
    })

    return homeValues
  }

  async fetchSpecies(people) {
    console.log(people)
    let species = people.map(person => {
      return fetchApi(person.species)
    })
    let allSpecies = await Promise.all(species)
    let speciesName = allSpecies.map(species => {
      return {
        name: species.name,
        language: species.language
      }
    })
    return speciesName;
  }

  render() {
    const { films, data } = this.state
    return (
      <div className="App">
        <Header />
        <ButtonsContainer fetchPeople={this.fetchPeople}/>
        <CardsContainer data={data}/>
        <FilmScroller films={films}/>
      </div>
    );
  }
}

export default App;
