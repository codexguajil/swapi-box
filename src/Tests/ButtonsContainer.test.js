import React from 'react';
import { ButtonsContainer } from '../Components/ButtonsContainer';
import { shallow } from 'enzyme';

describe("ButtonsContainer", () => {

    let wrapper = shallow (
        <ButtonsContainer />
    )

    it('should render some people, planet, and vehicles buttons', () => {
        expect(wrapper).toMatchSnapshot()
    })
})