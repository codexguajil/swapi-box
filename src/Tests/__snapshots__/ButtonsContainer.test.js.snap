// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonsContainer should render some people, planet, and vehicles buttons 1`] = `
<div
  id="buttons-container"
>
  <Button
    value="people"
  />
  <Button
    value="planets"
  />
  <Button
    value="vehicles"
  />
</div>
`;
